File: mle_normal_intro.r
# R code to draw samples, plot histogram, and overlay density functions

# 1. Draw 100 samples from N(2,1)
samples <- rnorm(100, mean = 2, sd = 1)

# 2. Plot the histogram of the sample
hist(samples, freq = FALSE, col = 'grey', main = 'Histogram with Overlayed Density Functions', xlab = 'Value', ylab = 'Density')

# 3. Overlay 5 different density functions from N(mu,1), where mu = 0, 1, 2, 3, 4
mus <- 0:4
colors <- c('blue', 'green', 'red', 'purple', 'orange')

# Loop through each mu to add a density function
for (i in 1:length(mus)) {
  curve(dnorm(x, mean = mus[i], sd = 1), add = TRUE, col = colors[i])
}

# Add a legend to the plot
legend('topright', legend = paste('N(', mus, ',1)', sep = ''), fill = colors)

File: mle_normal_intro.py
# Python code to perform the tasks using numpy and matplotlib

import numpy as np
import matplotlib.pyplot as plt

# 1. Draw 100 samples from N(2,1)
samples = np.random.normal(2, 1, 100)

# 2. Plot the histogram of the sample
plt.hist(samples, bins=10, alpha=0.5, density=True, label='Sample Histogram')

# 3. Overlay 5 different density functions from N(mu,1), where mu = 0, 1, 2, 3, 4
mus = [0, 1, 2, 3, 4]
colors = ['blue', 'green', 'red', 'purple', 'orange']

for mu, color in zip(mus, colors):
    # Create a range of values from -3 to +7 (covering the significant range for all distributions)
    x = np.linspace(-3, 7, 200)
    # Calculate the normal distribution's PDF values
    y = (1 / (np.sqrt(2 * np.pi))) * np.exp(-0.5 * (x - mu)**2)
    # Plot the density function
    plt.plot(x, y, label=f'N({mu},1)', color=color)

# Show the legend
plt.legend()

# Show the plot
plt.show()
